### PR TITLE
feat(pytest): Add `it.only` equivalent to pytest

### PR DESCRIPTION
**Pull request by lobsterkatie**
This adds a custom marker to pytest which mimics the [functionality of jest's `it.only`](https://jestjs.io/docs/api#testonlyname-fn-timeout). When a test is decorated with the `@pytest.mark.only` decorator, it will cause all other tests to be skipped. 

Though this works across the codebase (in other words, marking one test as an `only` will cause the other 16K tests in the repo to be skipped if a bare `pytest` command is used), it works better (a.k.a., way faster) when tests are already being restricted to a particular file (in other words, if pytest is being run as `pytest path/to/some/test/file.py`).

It also adds a flake8 rule to ensure that no instances of the new marker make it into production.

### Screenshots:

The marker in use:

image

The flake8 rule triggering in the linter:

image

The flake8 rule triggering in the pre-commit hook:

image

### CODE REVIEW

**Comment by lobsterkatie** [comment]
> the intended mechanism for this is to use `pytest`'s `-k` argument -- or run the test by fully qualified id (or `-m` if it's specifically a mark)
> 
> by not putting the mark in pyproject.toml it is an error already so no need for flake8
> 
> I'm -1 on this, there's already intended pytest mechanisms for doing what you want

Yeah, I know about those ways, but to me they feel less ergonomic. Using `-k` means a) I actually have to type out enough of the name of the test to be unique (which is both more work and more thinking than just typing the same trigger every time for the decorator snippet) and b) creates noise which makes it harder to find the test command in my terminal history should I want to go back to it. As for `-m`, yeah, I can make an alias for `pytest -m "only"`, but then I have to remember to switch to that command... None of this is insurmountable, of course, but I figured it might be nice for other folks who work in both JS and Python to have this one bit of feature parity between jest and pytest, and that if I wanted it to "just work" then others might, too.

> I'm already annoyed by how many mutually incompatible ways there are to write python tests.

Strike the word "Python" from that sentence and I'm totally with you - it's why I did this in the first place, trying to smooth over one of the differences between writing Python tests and JS tests. 

Bottom line, I did this for myself, but thought it would be at best nice to share with others and at worst neutral since it's purely additive (nothing changes for anyone not using it). If you'd rather close it, though, that's fine - I can just make the changes locally and manually gitignore them.

**Comment by untitaker** [comment]
@lobsterkatie do you know about `pytest -f`? It watches files and reruns failing tests for you

**Comment by lobsterkatie** [comment]
> @lobsterkatie do you know about `pytest -f`? It watches files and reruns failing tests for you

I do, but about half the time it tells me nothing failed last time and so it's going to run everything.

**Comment by evanpurkhiser** [comment]
> Yeah, I know about those ways, but to me they feel less ergonomic. Using -k means a) I actually have to type out enough of the name of the test to be unique (which is both more work and more thinking than just typing the same trigger every time for the decorator snippet) and b) creates noise which makes it harder to find the test command in my terminal history should I want to go back to it. As for -m, yeah, I can make an alias for pytest -m "only", but then I have to remember to switch to that command... None of this is insurmountable, of course, but I figured it might be nice for other folks who work in both JS and Python to have this one bit of feature parity between jest and pytest, and that if I wanted it to "just work" then others might, too.

I very much agree with this.

Using `.only` in jest tests is something I am constantly doing, and really keeps me in the flow of fixing tests that are broken one by one and not having to do a bunch of guessing with what I need to put for the `-k` option. Sure I can always be exact and put `ClassName::test_my_thing` but I would much prefer just slapping a `@pytest.only` on top of the test.

**Comment by mattgauntseo-sentry** [comment]
What happens if someone accidentally commits a .only to master? Is there a way to tell CI to ignore it?

**Comment by wedamija** [comment]
I'm generally -1 on this as well but I don't think it really hurts to merge it, since we have the lint rule in place to stop people merging it in their prs. It'd be even better if there was a way to tell CI to ignore it like @mattgauntseo-sentry suggested.

**Comment by lobsterkatie** [comment]
> Is there a way to tell CI to ignore it?

Yup. Done.

**Comment by lobsterkatie** [comment]
So what's the final verdict on this? As evidenced by my experience and Evan's comment above, I do think this is a win for folks who are used to Jest (and a non-issue for anyone else, as nothing says anyone has to use it).

**Comment by asottile-sentry** [comment]
I don't want to be the final verdict on this -- but (and especially as a pytest dev) I think we should not reimplement pytest features in different ways and should just use the standard ways to do things.  it feels like "just another feature" but it's something the dev-infra team would have to maintain and continue to adapt indefinitely
